---
title: "Preprocess cohort data"
output: html_notebook
---

```{r}
library(pheatmap)
library(GenomicRanges)
```

## Cohort data preprocessing
We first preprocess the data for the individual cohorts and save it in a single
RData file for more convenient analysis later on.
We remove any ranges from the ranges collection for each sentinel which had no
data available (e.g. genes having no epxression probes in the data).

```{r}
# parse the sentinels from the available data files
files <- list.files("data/current/cohorts/", ".*ranges.RData", full.names = F)
sentinels <- unlist(lapply(sapply(files, strsplit, "\\."), "[[", 1))
# get data for both cohorts for all sentinels
cohorts <- c("kora", "lolipop")

# for plotting the expression plots
plot.dir <- "results/current/plots/expression_per_sentinel"
if(!dir.exists(plot.dir)){
  dir.create(plot.dir, recursive = T)
}
data <- lapply(sentinels, function(sentinel){
  d <- lapply(cohorts, function(co){
    # load data into new environment
    env <- new.env()
    dfile <- paste0("data/current/cohorts/", 
                    sentinel, ".", toupper(co), ".adjusted.data.RData")
    if(!file.exists(dfile)){
      warning(paste0(co, " data does not exist for sentinel ", sentinel, "."))
    } else {
      load(dfile, envir=env)
      ggm.data <- with(env, processed)
      nodes <- colnames(ggm.data)
      # adjust ranges for only those which were initially found in the data
      load(paste0("data/current/cohorts/", 
                  sentinel, ".ranges.RData"))
      ranges$cpg.genes <- ranges$cpg.genes[(ranges$cpg.genes$SYMBOL) %in% nodes]
      ranges$snp.genes <- ranges$snp.genes[(ranges$snp.genes$SYMBOL) %in% nodes]
      ranges$spath <- ranges$spath[ranges$spath$SYMBOL %in% nodes]
      ranges$tfs <- ranges$tfs[ranges$tfs$SYMBOL %in% nodes]
      return(list(id=id, data=ggm.data, ranges=ranges, nodes=nodes))
    }
  })
  names(d) <- cohorts
  return(d)
})
names(data) <- sentinels
save(file="results/current/data.processed.RData", data)
```

In addition to the adjustment of the ranges, we create expression correlation plots
for all genes in a single locus and save them for later.
We also report any loci, which show a mean absolute correlation of $\rho > 0.05$.

```{r}
hmps <- unlist(lapply(names(data), function(sentinel){
  d <- data[[sentinel]]
  lapply(cohorts, function(co){
    expr.data <- d[[co]]$data[,!grepl("^rs|^cg", d[[co]]$nodes)]
    cor.mat <- cor((expr.data))
    mean.cor <- mean(cor.mat)
    if(abs(mean.cor) >= 0.05){
      cat(sentinel, "\t", co, "\t", mean.cor, "\n")
    }
    # plot gene expression correlations
    pdf(paste0(plot.dir, "/", co, ".", sentinel,".pdf"))
    pheatmap(cor.mat)
    dev.off()
  })
}))
```


## eQTL preprocessing
We use the eQTL results from LOLIPOP and KORA to validate our cis genes (cis-eQTL)
and the tfs/sp/cpg genes (trans-eQTL) identifiied using the GGM. 
We got a list from B. Lehne which contains the betas and standard errors for each
sentinel and the corresponding probes. We get the pvalue from the t-distribution 
by using n-2 degrees of freedom.


```{r}
# get eqtl results
env <- new.env()
load("data/current/cohorts/eqtls/lolipop-eqtls.RData", envir = env)
eqtl <- with(env, results)
rm(env)

# select the relevant cohorts (only oX_ia for lolipop)
eqtl <- eqtl[,c("snp", "probe", "beta.oX_ia", "se.oX_ia", "beta.kora", "se.kora")]

# get the t-statistic
eqtl <- with(eqtl, cbind(eqtl, stat.oX_ia=beta.oX_ia/se.oX_ia, stat.kora=beta.kora/se.kora))

# get the pvals and corrected pvals
eqtl <- with(eqtl, cbind(eqtl, 
                         p.lolipop = 2*pnorm(-abs(stat.oX_ia)), 
                         p.kora  = 2*pnorm(-abs(stat.kora))))
eqtl <- with(eqtl, cbind(eqtl, 
             q.lolipop = p.adjust(p.lolipop, "BH"),
             q.kora  = p.adjust(p.kora, "BH")))

# save the processed matrix, only keep p-vals and qvals
eqtl <- eqtl[,c("snp", "probe", "p.lolipop", "q.lolipop", "p.kora", "q.kora")]

save(file="results/current/eqtls.RData", eqtl)

```

