---
title: "Additional evaluation of simulation results spawned from reviewer comments"
output: 
 html_notebook:
    toc: TRUE
---

```{r setup}
library(dplyr)
library(readr)
library(grid)
library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(ggpubr)
library(scales)
library(reshape2)
library(circlize)
library(ggbio)
library(xtable)

# get the hg19 chromosome definitions
library(BSgenome.Hsapiens.UCSC.hg19)
hg19info <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)

# set up theme and colors
theme_set(theme_cowplot() + background_grid())
theme_update(legend.text = element_text(size=11), 
             legend.title=element_text(size=12),
             axis.text.x = element_text(size=10),
             axis.text.y = element_text(size=10))

# manually define the colors for the different methods (to get paired ones for
# prior/non prior versions)
paired <- brewer.pal(4, "Paired")
names(paired) <- c("glasso", "glasso (priors)", "bdgraph", "bdgraph (priors)")
#paired <- c(paired, "#66D35F")
#names(paired) <- c("glasso", "glasso (priors)", 
#                   "bdgraph (empty)", "bdgraph (priors)", "bdgraph (full)")
unpaired <- brewer.pal(7, "Dark2")[c(2,3,7)]
names(unpaired) <- c("irafnet", "genie3", "genenet")
graph_cols <- c(paired, unpaired)

sfb_graphs <- scale_fill_manual(values=graph_cols)
#sfb_graphs <- scale_fill_brewer(palette="Set2")
#scb_graphs <- scale_color_brewer(palette="Set2")
scb_graphs <- scale_color_manual(values=graph_cols)
sfb_binary <- scale_fill_brewer(palette = "Accent")
scb_binary <- scale_color_brewer(palette = "Accent")
scb_priors <- scale_color_brewer(palette = "Dark2")

bgm <- background_grid(major = "xy")
group_cols <- brewer.pal("Set2", n=3)
COLORS <- list(MEQTL = group_cols[1],
               EQTL = group_cols[2])

```

## Minimal subset analyis

Here we plot the progression of performance for increasing noise in the priors
for the 'minimal subset' of data, i.e. for a total of only 70 samples.

```{r minimal_subset}
# minimal subset size, prior noise progression
res <- read_tsv("results/current/biogrid_stringent/simulation/validation-subsetminimal.txt") %>%
  mutate(R = paste0("R=", rdegree))

# create nicer method names
tab <- res %>% 
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method=comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))

# get the MCC plot
simulation_mcc <- ggplot(tab,
                         aes(y = MCC,
                             x = R,
                             color = method)) +
  stat_boxplot(geom = "errorbar", width = .75) +
  geom_boxplot(
    outlier.size = 0,
    alpha = 0.5,
    coef = 0,
    outlier.shape = NA,
    
  ) +
  stat_summary(
    fun.y = median,
    geom = "smooth",
    position = position_dodge(0.75),
    aes(group = method),
    lwd = 0.8
  ) +
  scb_graphs +
  background_grid("none") +
  geom_vline(
    xintercept = 11.5,
    size = 1.5,
    color = "black",
    linetype = "dashed"
  ) +
  labs(x = "prior error",
       y = "MCC",
       fill = "") +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 12
    ),
    axis.title.x = element_text(size = 14,
                                margin = margin(-1, 0, 0, 0, unit = "lines")),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  )
simulation_mcc
```

For the response we combine the above plot with the original from the MS:

```{r}
finput <- paste0("results/current/biogrid_stringent/simulation/validation-subsetall.txt")

# create data-matrix
res <- read_tsv(finput) %>%
  mutate(R = paste0("R=", rdegree))


# create nicer method names
tab <- res %>% 
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  dplyr::rename(method=comparison)
toplot <- tab %>%
  filter(method != "bdgraph (full)") %>%
  mutate(is_prior_based = method %in% c("bdgraph (priors)", 
                                        "glasso (priors)",
                                        "irafnet")) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))
  
# get the MCC plot
simulation_mcc_orig <- ggplot(toplot,
                         aes(y=MCC, 
                             x=R, 
                             color=method)) +
  stat_boxplot(geom="errorbar", width=.75)+
  geom_boxplot(outlier.size=0, alpha=0.5, coef=0, outlier.shape = NA, ) + 
  stat_summary(fun.y=median, geom="smooth", 
               position=position_dodge(0.75),
               aes(group=method),lwd=0.8) +
  scb_graphs +
  #geom_boxplot(data = tab, alpha=0.5, aes(y=density_true, x=R), 
  #             fill="#666666", color="#666666",
  #             inherit.aes = FALSE,
  #             width=.3) +
  #scale_y_continuous(limits=c(min(tab$MCC),1), 
  #                   breaks=c(0,0.25,0.5,0.75,1),
  #                   sec.axis = sec_axis(trans = ~ ., 
  #                                       name="true graph density",
  #                                       breaks=seq(0,0.7,by=0.1))) +
  background_grid(major="x") +
  geom_vline(xintercept = 11.5, size=1.5, color="black", linetype="dashed") + 
  labs(x="prior error",
       y="MCC",
       fill="") + 
  theme(legend.position = "right",
        legend.text = element_text(size=12),
        legend.title = element_text(size=14), 
        axis.text.x = element_text(hjust=0, vjust=0.5, angle=-45, size=12),
        axis.title.x = element_text(size=14, 
                                    margin = margin(-1, 0, 0, 0, unit = "lines")),
        plot.margin = margin(0.2,1,0.2,0.2,"cm"))

ga <- ggarrange(simulation_mcc, simulation_mcc_orig, nrow = 2, 
                labels = "AUTO", common.legend = T, legend = "right")

ga

save_plot("results/current/revisions/figures/simulation_minimal_vs_maximial_subset_size.pdf",
          ga, nrow = 1.8, ncol = 1.6)
```

## Low /high noise scenario over all subset sizes

Here we plot the performance for a low and high noise scenario (0.1 and 0.7) for all sample subset sizes available.

```{r low_high_noise_all_subsets}
# minimal subset size, prior noise progression
res <- read_tsv("results/current/biogrid_stringent/simulation/validation-subsets.txt") %>%
  mutate(R = paste0("R=", rdegree)) %>% 
  filter(R %in% "R=0.8")

# create nicer method names
tab <- res %>% 
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method=comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))

# sample subset size plot with high noise level
simulation_sample_size_mcc <- ggplot(tab,
                                     aes(y = MCC,
                                         x = factor(subset),
                                         color = method)) +
  scb_graphs +
  stat_boxplot(geom = "errorbar", width = .75) +
  geom_boxplot(
    outlier.size = 0,
    alpha = 0.5,
    coef = 0,
    outlier.shape = NA
  ) +
  stat_summary(
    fun.y = median,
    geom = "smooth",
    position = position_dodge(0.75),
    aes(group = method),
    lwd = 0.8
  ) +
  scale_y_continuous(
    limits = c(min(tab$MCC), 1),
    breaks = c(0, 0.25, 0.5, 0.75, 1)
    
  ) +
  labs(color = "method",
       x = "subset size") + 
  theme(axis.text.x = element_text(hjust=0, vjust=0.5, angle=-45, size=12))

simulation_sample_size_mcc

#save_plot("results/current/figures/thesis/simulation_performance_sample_size_thesis.pdf",
#          simulation_sample_size_mcc + theme(legend.position = "bottom"), base_asp = 2)

```
Caption: Analysis of the effect of sample subset size on inference performance in a high prior noise setting (R=0.8).
Y-axis shows the MCC of inferred graphs VS the simulated ground truth, x-axis indicates the different sample sizes for inference increasing from left to right. Colors indicate the different models. 50 iterations of sampling were performed. Boxplots show medians (horizontal line) and first and third quartiles (lower/upper box borders). Whiskers show 1.5*IQR (inter-quartile range).

## Prior completeness

Here we plot the progression of method performance for dropping an increasing amount of information from the prior matrix (10%, 20%,... 90%).

```{r prior_completeness}
# minimal subset size, prior noise progression
res <- read_tsv("results/current/biogrid_stringent/simulation/validation-prior-completeness.txt") %>%
  mutate(fraction = paste0("F=", gsub("fraction", "", fraction_to_keep)))

# create nicer method names
tab <- res %>% 
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method=comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))

# get the MCC plot
simulation_mcc <- ggplot(tab,
                         aes(y = MCC,
                             x = fraction,
                             color = method)) +
  stat_boxplot(geom = "errorbar", width = .75) +
  geom_boxplot(
    outlier.size = 0,
    alpha = 0.5,
    coef = 0,
    outlier.shape = NA,
    
  ) +
  stat_summary(
    fun.y = median,
    geom = "smooth",
    position = position_dodge(0.75),
    aes(group = method),
    lwd = 0.8
  ) +
  scb_graphs +
  background_grid("none") +
  labs(x = "fraction of prior information kept",
       y = "MCC",
       fill = "") +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 12
    ),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  )

simulation_mcc
# save_plot(file="results/current/revisions/figures/simulation_prior_completeness.pdf",
#           simulation_mcc,
#           ncol = 1.6,
#           nrow = 1.2)
```
Caption: Analysis of the effect of 'prior completeness' on inference performance.
Y-axis shows the MCC of inferred graphs VS the simulated ground truth, x-axis indicates the fraction
of prior information retained for inference, increasing from left to right. Colors indicate the different models. Boxplots show medians (horizontal line) and first and third quartiles (lower/upper box borders). Whiskers show 1.5*IQR (inter-quartile range).

```{r}
print(length(unique(tab$snp)))
print(length(unique(tab$iteration)))
```

```{r prior_completeness_stat}
#' method: e.g. "glasso" or "bdgraph"
#' fraction: e.g. "F=0.1"
compare_wilcoxon <- function(method, fraction) {
  print(paste0("Results for method '", method, 
               "' and fraction '", fraction, "':"))
  
  tab_for_stat <- tab %>%
    filter(grepl(!!method, method)) %>%
    filter(fraction %in% !!fraction) %>% 
    select(method, MCC) %>%
    reshape2::melt()
  
  print("Models in comparison:")
  print(paste0(unique(tab_for_stat$method), collapse = ","))
  print("Results for wilcoxon test:")
  print(wilcox.test(value ~ method, data = tab_for_stat, 
                    alternative = "two.sided"))
}

compare_wilcoxon("glasso", "F=0.1")
compare_wilcoxon("bdgraph", "F=0.1")
```

## Prior completeness and noise

Here we plot the progression of method performance for prior comleteness and prior noise simulateneously.

```{r prior_completeness_and_noise}
# prior_completeness
completeness <- read_tsv("results/current/biogrid_stringent/simulation/validation-prior-completeness.txt") %>%
  mutate(completeness = as.numeric(gsub("fraction", "", fraction_to_keep))) %>%
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method=comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))

noise <- read_tsv("results/current/biogrid_stringent/simulation/validation-subsetall.txt") %>%
  filter(rdegree != "rbinom") %>%
  mutate(noise = as.numeric(rdegree)) %>%
  mutate(iteration = as.numeric(gsub("-.*", "", iteration))) %>%
  mutate(comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
         comparison = gsub("glasso$", "glasso (priors)", comparison),
         comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
         comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
         comparison = gsub("glasso_no_priors","glasso", comparison)) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method=comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))


toplot <-
  bind_rows(
    select(completeness, value = completeness, method, iteration, MCC) %>%
      mutate(variable = "missingness", 
             value = 1-value),
    select(noise, value = noise, method, iteration, MCC) %>%
      mutate(variable = "error")
  )


greens <- RColorBrewer::brewer.pal(3, "Greens")
blues <- RColorBrewer::brewer.pal(3, "Blues")

# get the MCC plot
noise_vs_completeness <-
  filter(toplot, method %in% c("bdgraph (priors)", "glasso (priors)")) %>%
  filter(value != 0 & value != 1) %>%
  mutate(group = paste0(method, " / ", variable)) %>%
  ggplot(aes(
    y = MCC,
    x = as.factor(value),
    color = group
  )) +
  stat_boxplot(geom = "errorbar", width = .75) +
  geom_boxplot(
    outlier.size = 0,
    alpha = 0.5,
    coef = 0,
    outlier.shape = NA,
    
  ) +
  stat_summary(
    fun.y = median,
    geom = "smooth",
    position = position_dodge(0.75),
    aes(group = group),
    lwd = 0.8
  ) +
  scale_color_manual(values = c("bdgraph (priors) / missingness" = greens[1],
                                "bdgraph (priors) / error" = greens[2],
                                "glasso (priors) / missingness" = blues[1],
                                "glasso (priors) / error" = blues[2])) +
  background_grid("none") +
  labs(x = "amount of error / missing priors",
       y = "MCC",
       color = "") +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 14
    ),
    plot.margin = margin(0.2, 0.5, 0.2, 0.5, "cm")
  )

noise_vs_completeness

cowplot::save_plot("simulation_completeness_vs_noise.pdf",
                   noise_vs_completeness, ncol = 1.7, nrow = 1.7)

```

## Sensititivy figure

```{r}
RESULT_PATH <- "results/current/biogrid_stringent/"
# generate the supplementary performance tables
finput <-
  paste0(RESULT_PATH, "simulation_rerun/validation-subsetall.txt")

priornoise_results <- read_tsv(finput) %>%
  mutate(R = paste0("R=", rdegree)) %>%
  mutate(
    comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
    comparison = gsub("glasso$", "glasso (priors)", comparison),
    comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
    comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
    comparison = gsub("glasso_no_priors", "glasso", comparison)
  ) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method = comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  ))

# sensititivity table as graph
sens_noise <- priornoise_results %>%
  group_by(method, R) %>%
  summarise(n = mean(`sensitivity`)) %>%
  ggplot(aes(x = factor(R), y = n, color = method)) +
  geom_point() +
  geom_line(aes(group = method)) +
  scb_graphs +
  cowplot::theme_cowplot() +
  cowplot::background_grid() +
  geom_vline(
    xintercept = 11.5,
    size = 1.5,
    color = "black",
    linetype = "dashed"
  ) +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 14
    ),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  ) +
labs(y = "",
     x = "prior error")

# also create graph for the sample size simulation
finput <- file.path(RESULT_PATH, "simulation_rerun/validation-subsets.tsv")
subsets_results <- read_tsv(finput) %>%
  mutate(
    comparison = gsub("bdgraph$", "bdgraph (priors)", comparison),
    comparison = gsub("glasso$", "glasso (priors)", comparison),
    comparison = gsub("bdgraph_no_priors$", "bdgraph (empty)", comparison),
    comparison = gsub("bdgraph_no_priors_full$", "bdgraph (full)", comparison),
    comparison = gsub("glasso_no_priors", "glasso", comparison)
  ) %>%
  filter(comparison != "bdgraph (full)") %>%
  dplyr::rename(method = comparison) %>%
  mutate(method = gsub("bdgraph \\(empty\\)", "bdgraph", method)) %>%
  mutate(method = factor(
    method,
    ordered = T,
    levels = c(
      "bdgraph",
      "bdgraph (priors)",
      "glasso",
      "glasso (priors)",
      "irafnet",
      "genie3",
      "genenet"
    )
  )) %>%
  mutate(subset = factor(subset, level = c(seq(50, 600, by = 50)), ordered = T))

# Sensitivity table
sens_subset <- subsets_results %>%
  group_by(method, subset) %>%
  summarise(n = mean(`sensitivity`)) %>%
  ggplot(aes(x = factor(subset), y = n, color = method)) +
  geom_point() +
  geom_line(aes(group = method)) +
  scb_graphs +
  cowplot::theme_cowplot() +
  cowplot::background_grid() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 14
    ),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  ) +
labs(y = "sensititvy",
     x = "subset size")

ga <- ggarrange(
    sens_subset,
    sens_noise,
    ncol = 2,
    legend = "right",
    common.legend = T, align = "h", labels = "AUTO"
) + 
  theme(plot.margin = margin(0,0.5,0,0,"cm"))

cowplot::save_plot("simulation_sensitivity.pdf",
                   ga, ncol = 2, nrow = 1)
```

## Specificity figure

```{r}
# specificity table as graph
spec_noise <- priornoise_results %>%
  group_by(method, R) %>%
  summarise(n = mean(`specificity`)) %>%
  ggplot(aes(x = factor(R), y = n, color = method)) +
  geom_point() +
  geom_line(aes(group = method)) +
  scb_graphs +
  cowplot::theme_cowplot() +
  cowplot::background_grid() +
  geom_vline(
    xintercept = 11.5,
    size = 1.5,
    color = "black",
    linetype = "dashed"
  ) +
  theme(
    legend.position = "right",
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 14
    ),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  ) +
labs(y = "",
     x = "prior error")

# specificity table as graph
spec_subset <- subsets_results %>%
  group_by(method, subset) %>%
  summarise(n = mean(`specificity`)) %>%
  ggplot(aes(x = factor(subset), y = n, color = method)) +
  geom_point() +
  geom_line(aes(group = method)) +
  scb_graphs +
  cowplot::theme_cowplot() +
  cowplot::background_grid() +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 14),
    axis.text.x = element_text(
      hjust = 0,
      vjust = 0.5,
      angle = -45,
      size = 14
    ),
    plot.margin = margin(0.2, 1, 0.2, 0.2, "cm")
  ) +
labs(y = "specificity",
     x = "subset size")

ga <- ggarrange(
    spec_subset,
    spec_noise,
    ncol = 2,
    legend = "right",
    common.legend = T, align = "h", labels = "AUTO"
) + 
  theme(plot.margin = margin(0,0.5,0,0,"cm"))

cowplot::save_plot("simulation_specificity.pdf",
                   ga, ncol = 2, nrow = 1)
```


## Sensitivity and specificty figures combined

```{r}
ga <- ggarrange(
  sens_subset + scale_y_continuous(limits = c(0,1)) + labs(x = ""),
  sens_noise + scale_y_continuous(limits = c(0,1)) + labs(x = ""),
  spec_subset + scale_y_continuous(limits = c(0,1)),
  spec_noise + scale_y_continuous(limits = c(0,1)),
  ncol = 2,
  nrow = 2,
  legend = "right",
  common.legend = T,
  align = "h",
  labels = "AUTO"
) +
  theme(plot.margin = margin(0, 0.5, 0, 0, "cm"))

cowplot::save_plot("simulation_sensitivity_specificity.pdf",
                   ga, ncol = 2, nrow = 2)
```
