#' -----------------------------------------------------------------------------
#' Perform fastENLOC analysis for the GTEx LBM locus in Skeletal Muscle
#'
#' @author Johann Hawe <johann.hawe@helmholtz-muenchen.de>
#'
#' @date Tue Mar 17 15:58:15 2020
#' -----------------------------------------------------------------------------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width=8, fig.height=7)
knitr::opts_knit$set(root.dir=normalizePath(".."))
knitr::opts_knit$set(tidy=TRUE)
library(tidyverse)
library(IRanges)
threads <- 6
```

Load the eQTL and GWAS results.
We get all trans genes for the 'rs9274623' SCZ locus.

```{r}
gwas <- read_tsv("data/current/gwas_atlas/lean_body_mass/wholebodyleanmass.results.metal_.txt") %>%
  select(SNP=MarkerName, Allele1, Allele2, Freq1, beta=Effect, P=`P-value`) %>%
  mutate(Allele1 = toupper(Allele1), Allele2 = toupper(Allele2)) %>%
  mutate(z = (-sign(beta) * qnorm(P / 2)))

# alternative GWAS
gwas <- read_tsv("data/current/gwas_summary_stats/FFM_c_maf0.01_meta_pos.txt") %>%
  select(SNP=MarkerName, Allele1, Allele2, Freq1, beta=Effect, P=Pvalue, pos, chr) %>%
  mutate(Allele1 = toupper(Allele1), Allele2 = toupper(Allele2)) %>%
  mutate(z = (-sign(beta) * qnorm(P / 2)))


# we set positions manually, as we had to convert the original gtex snp location
# to the hg19 position
# gtex trans eQTL: https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_eGenes_fdr05.txt
# gtex snp info: https://www.gtexportal.org/home/snp/rs9318186
# snpID: rs9318186


# list of genes we want to analyze (compare trans eQTLs above)
trans_genes <- c("SYNC", "PHOSPHO1", "FAM109A", "CACNG1", "CSPG4")

trans_eqtl <- read_tsv("https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_trans_eGenes_fdr05.txt") %>% filter(tissue_id == "Muscle_Skeletal" & gene_name %in% trans_genes) %>%
   mutate(z = (-sign(slope) * qnorm(pval_nominal / 2)))

locus <- "rs9318186"
snp_chr <- 13
cis_snp_pos <- 74106939

# the gene is KLF5
cis_gene <- "KLF5"

trans_snps_pos <- c(74156765, 74115700, 74106939 ,74108587)
trans_snps_rs <- c("rs60984169", "rs7326612", "rs9318186", "rs1028883")
trans_snps_id <-  c("chr13_73582628_A_ACT_b38",
                    "chr13_73541563_T_C_b38", 
                    "chr13_73532802_A_G_b38", 
                    "chr13_73534450_T_G_b38")
trans_snps <- tibble(id=trans_snps_id,
                     rsid=trans_snps_rs,
                     pos=trans_snps_pos,
                     chr=snp_chr)
region_start <- min(trans_snps$pos)
region_end <- max(trans_snps$pos)

trans_eqtl <- left_join(trans_eqtl, trans_snps, by=c("variant_id" = "id"))

# get subset of data to look at
gwas_roi <- filter(gwas, SNP %in% trans_snps$rsid) %>%
  left_join(trans_snps, by=c("SNP" = "rsid"))

# filter trans-eQTL only for our region of interest
eqtl_trans_roi <- filter(eqtl_trans, 
                         SNPChr == snp_chr & 
                           SNPPos >= region_start & 
                           SNPPos <= region_end)

```

Define the output directory and write for each gene individually the respective 
GWAS and eQTL information including the list of SNPs to be analyzed

```{r}
dout <- "results/current/colocalization_lbm/"
dir.create(dout)

write_tsv(trans_eqtl, path = file.path(dout, "eqtl.tsv"))
write_tsv(trans_eqtl %>% select(rsid), 
            path = file.path(dout, "eqtl_snps.txt"), col_names = F)
write_tsv(gwas_roi, path = file.path(dout, "lbm.tsv"))

```


```{bash, eval=FALSE}
dir=results/current/colocalization_lbm/
/home/icb/johann.hawe/tools/plink/plink --bfile /storage/groups/epigenereg01/workspace/public_data/polygenic_risk_scores/1000genomes/riskSNPs_AF_CAD_EUR --extract $dir/eqtl_snps.txt --freq --r square --out $dir/eqtl_snps_1kg_EUR &> $dir/plink.out
```

Final preparations comming up: Load eQTL, GWAS and 1kG data again, match common
snps, prepare LD block information.

```{r}

dgene <- paste0("results/current/colocalization_lbm/") 
feqtl <- file.path(dgene, "eqtl.tsv")
if(!file.exists(feqtl)) {
  return(invisible(NULL))
}

print(dgene)

eqtl <- read_tsv(feqtl)
lbm <- read_tsv(file.path(dgene, "lbm.tsv"))
ld <- as.matrix(read.table(file.path(dgene, "eqtl_snps_1kg_EUR.ld")))
if(nrow(ld) < 2) return(NULL)

freq <- read.table(file.path(dgene, "eqtl_snps_1kg_EUR.frq"), header=T,
                   stringsAsFactors = F)
common_snps <- intersect(freq$SNP, eqtl$rsid)
common_snps <- intersect(common_snps, lbm$SNP)

freq_matched <- freq[freq$SNP %in% common_snps,]
ld_matched <- ld[freq$SNP %in% common_snps, freq$SNP %in% common_snps]
colnames(ld_matched) <- rownames(ld_matched) <- freq_matched$SNP

write.table(ld_matched, file=file.path(dgene, "eqtl_snps_1kg_EUR_common.ld"), 
            sep="\t", quote=F, row.names = FALSE, col.names = FALSE)

eqtl_matched <- eqtl[match(freq_matched$SNP, eqtl$rsid),] %>%
  mutate(chr = snp_chr)
lbm_matched <- lbm[match(freq_matched$SNP, lbm$SNP),]

write.table(eqtl_matched[,c("rsid", "z")], file.path(dgene, "eqtl_z.tsv"), 
            sep="\t", quote=F, row.names=F, col.names=F)

# chromosome to be loaded for LD block information
blocks <- read.table(paste0("/storage/groups/epigenereg01/workspace/public_data/ldetect-data/EUR/fourier_ls-chr", snp_chr, ".bed"), 
                     header=TRUE, stringsAsFactors = FALSE)
blocks <- IRanges(blocks$start, blocks$stop)
snp_gr <- IRanges(lbm_matched$pos, width=1)
snp2block <- subjectHits(findOverlaps(snp_gr, blocks))

## put togehter the GWAS
scz_LDblocks_z <- data.frame(SNP=lbm_matched$SNP,
                             LDblock=snp2block,
                             z=lbm_matched$z)

fname <- file.path(dgene, "lbm_gwas_zscores.txt")
write.table(scz_LDblocks_z, file=fname, sep="\t", 
            row.names = FALSE, col.names = FALSE, quote=F)
system(paste("gzip -f", fname))


# Finally, we have to summarize the dap results so they can be used in fastenloc. 
# We need a vcf file to annotate the SNPs. Most info is in the eQTL table.
lookup <- c('#CHROM'="chr",
  POS="pos",
  ID="rsid",
  REF=NA,
  ALT=NA,
  QUAL=NA,
  FILTER=NA,
  INFO=NA)
vcf <- eqtl_matched[,setdiff(lookup, NA)]
vcf <- cbind(vcf, matrix(".", nrow=nrow(vcf), ncol=sum(is.na(lookup))))
colnames(vcf) <- names(lookup)
write.table(vcf, file=file.path(dgene, "eqtl_snps.vcf"), 
            sep="\t", quote=F, row.names=F)

```

Now call DAP to get posterior probs and prepare the final input for fastENLOC

```{bash, eval=FALSE}
threads=12

  echo "Processing $(basename $dir)."
  
  ./packages/dap/dap_src/dap-g.static -t $threads -d_z $dir/eqtl_z.tsv -d_ld $dir/eqtl_snps_1kg_EUR_common.ld \
    --all -o $dir/eqtl_dap &> $dir/eqtl_dap_debug.out
  echo "dap-g done."
  # dap might fail due to unknown errors...
  success=`wc -l $dir/eqtl_dap | cut -d " " -f 1`
  if [ $success -eq "0" ] ; then
    continue
  fi
  
  gzip -f $dir/eqtl_snps.vcf
  mkdir -p $dir/dap_rst/
  ln -s ../eqtl_dap $dir/dap_rst/$(basename $dir).dap
  perl ./packages/fastenloc/src/summarize_dap2enloc.pl -dir $dir/dap_rst/ \
    -vcf $dir/eqtl_snps.vcf.gz -tissue eQTLgen | \
    gzip - > $dir/fastenloc.eqtl.annotation.vcf.gz
  ./packages/dap/torus_src/torus.static -d $g/scz_gwas_zscores.txt.gz --load_zval \
    -dump_pip $g/scz_gwas_pip.txt
  echo "torus done."
  gzip -f $g/scz_gwas_pip.txt
  total=`wc -l $g/scz.tsv | cut -d " " -f 1`
  ./packages/fastenloc/src/fastenloc -eqtl $g/fastenloc.eqtl.annotation.vcf.gz \
  -gwas $g/scz_gwas_pip.txt.gz \
  -total_variants $total \
  -t eQTLgen \
  -thread $threads \
  -prefix $g/fastenloc &> $g/fasentloc.debug.out
  
   echo "fastENLOC done."
   
done
```

Once the above bash script has been run, we can gather up the results and 
visualize them. We first load some additional plotting packages

```{r}
library(cowplot)
library(ggrepel)
library(ggpubr)
theme_set(theme_cowplot())
```

```{r}
data <- lapply(genes, function(gene) {
  dgene <- paste0("results/current/colocalization_scz/", gene) 
  ffastenloc <- file.path(dgene, "fastenloc.enloc.snp.out")
  if(!file.exists(ffastenloc)) {
    return(invisible(NULL))
  }
  eqtl <- read_tsv(file.path(dgene, "eqtl.tsv"))
  coloc_eqtl <- read.table(ffastenloc, stringsAsFactors = FALSE) %>% as_tibble()
  colnames(coloc_eqtl) <- c("signal.id", "snp", "QTL.PIP", "GWAS.PIP", "GWAS.PIP.QTL", "SCP")
  coloc_eqtl <- mutate(coloc_eqtl,
                       pos = eqtl[match(coloc_eqtl$snp, eqtl$SNP), "SNPPos",drop=T],
                       trait=sapply(strsplit(signal.id, ":"), "[[", 1))
  
  coloc_eqtl  

}) %>% bind_rows()

selected_trans <- c("CD6", "B9D2", "PSEN1", "DEF8", 
                    "DNAJB2", "ZNF672", "SH3BGRL3")


best <- data %>% filter(trait %in% selected_trans) %>%
  group_by(snp) %>% 
  summarise(maxSCP=max(SCP), 
            meanSCP = mean(SCP),
            pos=pos[which.max(SCP)],) %>% 
  top_n(3,meanSCP)

gp <- data %>% filter(trait %in% selected_trans) %>%
  ggplot(aes(x=pos, y=SCP,group=snp)) + 
  stat_summary(fun.y="mean", geom="point") +
#  stat_summary(fun.data = "mean_se", geom = "errorbar") +
  scale_y_continuous(limits=c(0,1)) + 
  geom_label_repel(aes(x=pos, y=(meanSCP), label=snp), 
                   data=best, segment.colour="black", box.padding = 1)
gp

# plot showing the gwas and eQTL results
toplot <- bind_rows(select(gwas_roi, snp=SNP, pos=BP, P=P) %>% mutate(trait="SCZ"),
                    select(eqtl_trans_roi, snp=SNP, pos=SNPPos, P=Pvalue) %>% mutate(trait="eQTL"))

toplot %>%
  ggplot(aes(x=pos, y=-log10(P), color=trait)) + 
  geom_point(alpha=0.5) + 
  facet_wrap(.~trait)

```
# ------------------------------------------------------------------------------
print("SessionInfo:")
# ------------------------------------------------------------------------------
sessionInfo()
