```{r setup, include=FALSE}
#' -----------------------------------------------------------------------------
#' fastENLOC analysis of the SCZ locus described in the GGM manuscript.
#' We perform two individual analyses, one for the trans genes one for the 
#' cis genes in the network.
#'
#' @author Johann Hawe <johann.hawe@helmholtz-muenchen.de>
#'
#' @date Tue Mar 17 15:58:15 2020
#' -----------------------------------------------------------------------------

knitr::opts_chunk$set(echo = TRUE, fig.width=8, fig.height=7)
knitr::opts_knit$set(root.dir=normalizePath(".."))
knitr::opts_knit$set(tidy=TRUE)
library(tidyverse)
library(IRanges)
library(cowplot)
theme_set(theme_cowplot() + background_grid(major="xy"))
threads <- 6
```

Load the eQTL and GWAS results.
We get all trans genes for the 'rs9274623' SCZ locus.

```{r message=F, error=F}

eqtl_trans_sign <- read_tsv("data/current/eqtl_gen/trans-eQTL_significant_20181017.txt.gz") 
eqtl_cis_sign <- read_tsv("data/current/eqtl_gen/cis-eQTL_significant_20181017.txt.gz")

locus <- "rs9274623"

snp_chr <- filter(eqtl_cis_sign, SNP == locus) %>% pull(SNPChr) %>% unique()
snp_pos <- filter(eqtl_cis_sign, SNP == locus) %>% pull(SNPPos) %>% unique()
extension_trans <- 500000
extension_cis <- 250000
region_start <- snp_pos - extension_trans
region_end <- snp_pos + extension_trans

# list of genes we want to analyze
genes_trans <- eqtl_trans_sign %>%
  filter(SNP == !!locus) %>% 
  pull(GeneSymbol)

# all cis genes in the sentinel region
# we use the original ranges collection for this
ranges <- readRDS(paste0("results/current/biogrid_stringent/ranges/", 
                         locus, 
                         "_eqtlgen.rds"))
genes_cis <- unique(ranges$snp_genes$SYMBOL)

# get GWAS results in the region of interest
gwas_roi <- read_tsv("data/current/gwas_atlas/schizophrenia/clozuk_pgc2.meta.sumstats.txt") %>%
  filter(CHR == !!snp_chr) %>%
  mutate(beta = log(OR)) %>%
  mutate(z = (- sign(beta) * qnorm(P / 2)))

# get the full list of associations for cis and trans
eqtl_trans_roi <- read_tsv("data/current/eqtl_gen/trans-eQTLs_full_20180905.txt.gz") %>%
  filter(GeneSymbol %in% genes_trans &
         SNPChr == snp_chr & 
         SNPPos >= region_start & 
         SNPPos <= region_end)
#eqtl_cis_roi <- read_tsv("data/current/eqtl_gen/cis-eQTLs_full_20180905.txt.gz") %>%
#  filter(GeneSymbol %in% genes_cis)

rm(ranges, eqtl_trans_sign, eqtl_cis_sign)

gc(full=T)

```

Define the output directory and write for each gene individually the respective 
GWAS and eQTL information including the list of SNPs to be analyzed

```{r}

# helper
process_genes <- function(genes, eqtl, dout, is_cis=F) {
  tmp <- mclapply(genes, function(gene) {
    eqtl_subset <- filter(eqtl, GeneSymbol == !!gene)
    if(is_cis) {
      # adjust window (trans would already have been filtered)
      gene_pos <- unique(pull(eqtl_subset, GenePos))
      gene_start <- gene_pos - extension_cis
      gene_stop <- gene_pos + extension_cis
      eqtl_subset <- filter(eqtl_subset, SNPPos > gene_start &
                              SNPPos < gene_stop)
    }
    if(nrow(eqtl_subset) > 1) {
        dout_gene <- file.path(dout, gene)
        dir.create(dout_gene)
        
        write_tsv(eqtl_subset, path = file.path(dout_gene, "eqtl.tsv"))
        write_tsv(eqtl_subset %>% select(SNP), 
                    path = file.path(dout_gene, "eqtl_snps.txt"), col_names = F)
        
        # gwas subset for the current gene, either cis region around locus SNP
        # for trans genes or around the cis gene itself
        if(is_cis) {
          gwas_subset <- filter(gwas_roi, BP>gene_start & BP<gene_stop)
        } else {
          gwas_subset <- filter(gwas_roi, BP>region_start & BP<region_end)  
        }
        
        write_tsv(gwas_subset, path = file.path(dout_gene, "scz.tsv"))
      }
    invisible(NULL)
  }, mc.cores = 1)
}

dout_cis <- "results/current/colocalization_scz/cis/"
dir.create(dout_cis, recursive = T)
process_genes(genes_cis, eqtl_cis_roi, dout_cis, is_cis = T)

dout_trans <- "results/current/colocalization_scz/trans/"
dir.create(dout_trans, recursive = T)
process_genes(genes_trans, eqtl_trans_roi, dout_trans)

```


```{bash, eval=FALSE}
for g in results/current/colocalization_scz/trans/* results/current/colocalization_scz/cis/* ; do
  echo "Processing $(basename $g)"
  /home/johann.hawe/tools/plink/plink \
  --bfile /storage/groups/epigenereg01/workspace/public_data/polygenic_risk_scores/1000genomes/riskSNPs_AF_CAD_EUR \
  --extract $g/eqtl_snps.txt \
  --freq \
  --r square \
  --out $g/eqtl_snps_1kg_EUR &> $g/plink.out 
done
```

Final preparations comming up: Load eQTL, GWAS and 1kG data again, match common
snps, prepare LD block information.

```{r}
process_final <- function(genes, dout) {
  tmp <- mclapply(genes, function(gene) {
   
    dgene <- file.path(dout, gene) 
    
    feqtl <- file.path(dgene, "eqtl.tsv")
    if(!file.exists(feqtl)) {
      return(invisible(NULL))
    }
    
    print(dgene)
    
    eqtl <- read_tsv(feqtl)
    scz <- read_tsv(file.path(dgene, "scz.tsv"))
    scz <- mutate(scz, rsid = sapply(strsplit(SNP, ":"), "[[", 1))
    ld <- as.matrix(read.table(file.path(dgene, "eqtl_snps_1kg_EUR.ld")))
    if(nrow(ld) < 2) return(NULL)
    
    freq <- read.table(file.path(dgene, "eqtl_snps_1kg_EUR.frq"), header=T,
                       stringsAsFactors = F)
    common_snps <- intersect(freq$SNP, eqtl$SNP)
    common_snps <- intersect(common_snps, scz$rsid)
    
    freq_matched <- freq[freq$SNP %in% common_snps,]
    ld_matched <- ld[freq$SNP %in% common_snps, freq$SNP %in% common_snps]
    colnames(ld_matched) <- rownames(ld_matched) <- freq_matched$SNP
    
    write.table(ld_matched, file=file.path(dgene, "eqtl_snps_1kg_EUR_common.ld"), 
                sep="\t", quote=F, row.names = FALSE, col.names = FALSE)
    
    eqtl_matched <- eqtl[match(freq_matched$SNP, eqtl$SNP),]
    scz_matched <- scz[match(freq_matched$SNP, scz$rsid),]
  
    ## flip the eQTL alleles to match the reference
    flip <- eqtl_matched$AssessedAllele == freq_matched$A2
    if(any(flip)) {
      tmp <- eqtl_matched$AssessedAllele[flip]
      eqtl_matched$AssessedAllele[flip] <- eqtl_matched$OtherAllele[flip]
      eqtl_matched$OtherAllele[flip] <- tmp
      eqtl_matched$Zscore[flip] <- -eqtl_matched$Zscore[flip]
    }
    write.table(eqtl_matched[,c("SNP", "Zscore")], file.path(dgene, "eqtl_z.tsv"), 
                sep="\t", quote=F, row.names=F, col.names=F)
    
    ## correlations are signed so we need to make sure that effects have a direction
    flip <- scz_matched$A1 == freq_matched$A2
    if(any(flip)) {
      tmp <- scz_matched$A1[flip]
      scz_matched$A1[flip] <- scz_matched$A2[flip]
      scz_matched$A2[flip] <- tmp
      scz_matched$z[flip] <- scz_matched$z[flip]
    }
    # chromosome to be loaded for LD block information
    chr <- unique(scz_matched$CHR)
    blocks <- read.table(paste0("/storage/groups/epigenereg01/workspace/public_data/ldetect-data/EUR/fourier_ls-chr", chr, ".bed"), 
                         header=TRUE, stringsAsFactors = FALSE)
    blocks <- IRanges(blocks$start, blocks$stop)
    snp_gr <- IRanges(scz_matched$BP, width=1)
    snp2block <- subjectHits(findOverlaps(snp_gr, blocks))
    
    ## put togehter the GWAS
    scz_LDblocks_z <- data.frame(SNP=scz_matched$rsid, LDblock=snp2block, z=scz_matched$z)
    
    fname <- file.path(dgene, "scz_gwas_zscores.txt")
    write.table(scz_LDblocks_z, file=fname, sep="\t", 
                row.names = FALSE, col.names = FALSE, quote=F)
    system(paste("gzip -f", fname))
    
    
    # Finally, we have to summarize the dap results so they can be used in fastenloc. 
    # We need a vcf file to annotate the SNPs. Most info is in the eQTL table.
    lookup <- c('#CHROM'="SNPChr",
      POS="SNPPos",
      ID="SNP",
      REF="OtherAllele",
      ALT="AssessedAllele",
      QUAL=NA,
      FILTER=NA,
      INFO=NA)
    vcf <- eqtl_matched[,setdiff(lookup, NA)]
    vcf <- cbind(vcf, matrix(".", nrow=nrow(vcf), ncol=sum(is.na(lookup))))
    colnames(vcf) <- names(lookup)
    write.table(vcf, file=file.path(dgene, "eqtl_snps.vcf"), 
                sep="\t", quote=F, row.names=F)
  
    invisible(NULL)
    
  }, mc.cores=1)
}

process_final(genes_cis, dout_cis)

process_final(genes_trans, dout_trans)


```

Now call DAP to get posterior probs and prepare the final input for fastENLOC

```{bash, eval=FALSE}
threads=6
# process both cis and trans prepped genes
for g in results/current/colocalization_scz/cis/* results/current/colocalization_scz/trans/* ; do
  # remove old results first
  rm -rf $g/dap_rst/
  rm -f $g/eqtl_dap $g/scz_gwas_pip.txt $g/scz_gwas_pip.txt.gz
  rm -f $g/fasten*
  
  if [ ! -e "$g/eqtl_snps.vcf" ] ; then continue ; fi
  
  echo "Processing $(basename $g)."
  
  ./packages/dap/dap_src/dap-g.static -t $threads -d_z $g/eqtl_z.tsv -d_ld $g/eqtl_snps_1kg_EUR_common.ld \
    --all -o $g/eqtl_dap &> $g/eqtl_dap_debug.out
  echo "dap-g done."
  
  # dap might fail due to unknown errors...
  success=`wc -l $g/eqtl_dap | cut -d " " -f 1`
  if [ $success -eq "0" ] ; then
    continue
  fi
  
  gzip -c -f $g/eqtl_snps.vcf > $g/eqtl_snps.vcf.gz
  mkdir -p $g/dap_rst/
  ln -s ../eqtl_dap $g/dap_rst/$(basename $g).dap
  perl ./packages/fastenloc/src/summarize_dap2enloc.pl -dir $g/dap_rst/ \
    -vcf $g/eqtl_snps.vcf.gz -tissue eQTLgen | \
    gzip - > $g/fastenloc.eqtl.annotation.vcf.gz
  
  success=`zcat $g/fastenloc.eqtl.annotation.vcf.gz | wc -l | cut -d " " -f 1`
  if [ $success -eq "0" ] ; then
    continue
  fi
  
  ./packages/dap/torus_src/torus.static -d $g/scz_gwas_zscores.txt.gz --load_zval \
    -dump_pip $g/scz_gwas_pip.txt &> $g/torus.debug.out
  echo "torus done."
  gzip -c -f $g/scz_gwas_pip.txt > $g/scz_gwas_pip.txt.gz
  
  # call fastENLOC 
  
  total=`wc -l $g/scz.tsv | cut -d " " -f 1`
  
  ./packages/fastenloc/src/fastenloc.static -eqtl $g/fastenloc.eqtl.annotation.vcf.gz \
  -gwas $g/scz_gwas_pip.txt.gz \
  -total_variants $total \
  -t eQTLgen \
  -thread $threads \
  -prefix $g/fastenloc &> $g/fastenloc.debug.out
  
   echo "fastENLOC done."
   
   # could happen we didn't get any results after all...
  success=`wc -l $g/fastenloc.enloc.snp.out | cut -d " " -f 1`
  if [ $success -eq "0" ] ; then
    rm $g/fastenloc.enloc*
  fi
   
done
```

```{r}
# function to get several LD plots for subsequent SNP ranges
ld_plots <- function(ld, size=1000) {
  
  require(reshape2)
  require(ggplot2)
  require(dplyr)
  
  
  myseq <- seq(1, ncol(ld), by=1000)
  
  lapply(2:(length(myseq)+1), function(seq) {
    if(seq == (length(myseq)+1)) {
      low <- myseq[seq-1]
      high <- ncol(ld)
    } else {
      low <- myseq[seq-1]
      high <- myseq[seq]
    }
    
    toplot <- melt(ld[low:high, low:high])
    toplot %>% ggplot(aes(x=Var1, y=Var2, fill=value)) + 
      geom_tile() + 
      scale_fill_gradient2(limits=c(-1,1)) + 
      theme(axis.text.x = element_blank(), 
            axis.text.y = element_blank())
  })
}

f <- file.path(dgene, "eqtl_snps_1kg_EUR.ld")
ld <- as.matrix(read.table(f))
plots <- ld_plots(ld)

```
Once the above bash script has been run, we can gather up the results and 
visualize them. We first load some additional plotting packages

```{r}
library(cowplot)
library(ggrepel)
library(ggpubr)
theme_set(theme_cowplot() + background_grid(major="xy"))
```

```{r}

get_results <- function(genes, directory) {
  lapply(genes, function(gene) {
    dgene <- paste0(directory, gene) 
    print(dgene)
    ffastenloc <- file.path(dgene, "fastenloc.enloc.snp.out")
    if(!file.exists(ffastenloc)) {
      return(invisible(NULL))
    }
    eqtl <- read_tsv(file.path(dgene, "eqtl.tsv"))
    coloc_eqtl <- read.table(ffastenloc, 
                             stringsAsFactors = FALSE, 
                            header = F) %>% as_tibble()
    colnames(coloc_eqtl) <- c("signal.id", "snp", "QTL.PIP", "GWAS.PIP", "GWAS.PIP.QTL", "SCP")
    coloc_eqtl <- mutate(coloc_eqtl,
                         pos = eqtl[match(coloc_eqtl$snp, eqtl$SNP), "SNPPos",drop=T],
                         trait=sapply(strsplit(signal.id, ":"), "[[", 1))
    
    coloc_eqtl  
  
  }) %>% bind_rows()
}

# get reusults for cis and trans individually
results_cis <- get_results(genes_cis, dout_cis)
results_trans <- get_results(genes_trans, dout_trans)

# save to file
write_tsv(results_cis, "results/current/colocalization_scz/results_cis.tsv")
write_tsv(results_trans, "results/current/colocalization_scz/results_trans.tsv")

# the trans genes selected in our network
selected_trans <- c("CD6", "B9D2", "PSEN1", "DEF8", 
                    "DNAJB2", "ZNF672", "SH3BGRL3", "CXCR5", "RNF114")

# cis genes directly connected to the SNP
selected_cis <- c("RNF5", "PBX2", "HLA-DQA1")

results_trans_filtered <- results_trans %>% filter(trait %in% selected_trans)
results_cis_filtered <- results_cis %>% filter(trait %in% selected_cis)

# plot the SCP plot for a specific data set
plot_scp <- function(d) {
  best <- d %>%
    top_n(3,SCP)
  gp <- d %>%
    ggplot(aes(x=pos, y=SCP)) + 
    geom_point(aes(color=trait)) + 
    #stat_summary(fun.data="mean_se", geom="errorbar") +
    scale_y_continuous(limits=c(0,1)) +
    geom_label_repel(aes(x=pos, y=SCP, label=paste0(snp,"|",trait)), 
                     data=best, segment.colour="black", box.padding = 1)
  gp
}

# create the full and the filtered plot for trans
gp_trans <- plot_scp(results_trans)
gp_trans_filtered <- plot_scp(results_trans_filtered)

# create the full and the filtered plot for cis
gp_cis <- plot_scp(results_cis)
gp_cis_filtered <- plot_scp(results_cis_filtered)

# overview plot
ggarrange(gp_trans, gp_trans_filtered, gp_cis, gp_cis_filtered, 
          ncol=2, nrow=2, labels="AUTO", align = "h")

# for the manuscript
final_trans <- gp_trans_filtered + labs(y = "SCP", 
                   x = "genomic position (chr 6)",
                   title = "Colocalization for trans-eQTL",
                   color = "gene")
final_cis <- gp_cis_filtered + labs(y = "mean SCP", 
                   x = "genomic position (chr 6)",
                   title = "Colocalization for cis-eQTL")
final_combined <- ggarrange(final_cis, final_trans, ncol=2, labels="AUTO", 
                            align="h")
save_plot("results/current/figures/scz_colocalization_trans.pdf", 
          final_trans, ncol=1, nrow=1)

# plot showing the gwas and trans eQTL results
toplot <- bind_rows(select(gwas_roi, snp=SNP, pos=BP, P=P) %>% mutate(trait="SCZ"),
                    select(eqtl_trans_roi, snp=SNP, pos=SNPPos, P=Pvalue) %>% mutate(trait="eQTL"))

toplot %>%
  ggplot(aes(x=pos, y=-log10(P), color=trait)) + 
  geom_point(alpha=0.5) + 
  scale_y_continuous(breaks=seq(0,160,by=10))
  #facet_wrap(.~trait)


# plot showing the gwas and cis eQTL results
toplot <- bind_rows(select(gwas_roi, snp=SNP, pos=BP, P=P) %>% mutate(trait="SCZ"),
                    select(eqtl_cis_roi, snp=SNP, pos=SNPPos, P=Pvalue) %>% mutate(trait="eQTL"))

toplot %>%
  ggplot(aes(x=pos, y=-log10(P), color=trait)) + 
  geom_point(alpha=0.5) + 
  scale_y_continuous(breaks=seq(0,160,by=10))+
  facet_wrap(.~trait)

```


```{r}
# ------------------------------------------------------------------------------
print("SessionInfo:")
# ------------------------------------------------------------------------------
sessionInfo()
```